libraries/AP_Scripting/examples/rangefinder_quality_test.lua

```diff
@@ -0,0 +1,248 @@
+
+local update_prepare
+local update_wait
+local update_begin_test
+local update_eval_test
+
+local function _update_prepare()
+    if Parameter('RNGFND1_TYPE'):get() ~= RNGFND_TYPE_LUA then
+        return complete("LUA range finder driver not enabled")
+    end
+    if rangefinder:num_sensors() < 1 then
+        return complete("LUA range finder driver not connected")
+    end
+    rngfnd_backend = rangefinder:get_backend(0)
+    if not rngfnd_backend then
+        return complete("Range Finder 1 does not exist")
+    end
+    if (rngfnd_backend:type() ~= RNGFND_TYPE_LUA) then
+        return complete("Range Finder 1 is not a LUA driver")
+    end
+
+    return update_wait()
+end
+
+local function _update_wait()
+    -- Check for timeout while initializing
+    if millis():tofloat() - time_start_ms > TIMEOUT_MS then
+        return complete("Timeout while trying to initialize")
+    end
+
+    -- Wait until the prearm check passes. This ensures that the system is mostly initialized
+    -- before starting the tests.
+    if not arming:pre_arm_checks() then
+        return update_wait, UPDATE_PERIOD_MS
+    end
+
+    -- Do some one time tests
+    local error_str = do_status_tests()
+    if error_str then
+        return complete(error_str)
+    end
+
+    -- Continue on to the main list of tests.
+    return update_begin_test()
+end
+
+local function _update_begin_test()
+    test_idx = test_idx + 1
+    if test_idx > #test_data then
+        return complete()
+    end
+
+    local dist_m_in = test_data[test_idx][1]
+    local signal_quality_pct_in = test_data[test_idx][2]
+
+    -- Use the driver interface to simulate a data measurement being received and being passed to AP.
+    local result = rngfnd_backend:handle_script_msg2(dist_m_in, signal_quality_pct_in)
+    if not result then
+        return complete(string.format("Test %i, dist_m: %.2f, quality_pct: %3i failed to handle_script_msg2",
+            test_idx, dist_m_in, signal_quality_pct_in))
+    end
+
+    return update_eval_test, UPDATE_PERIOD_MS
+end
+
+local function _update_eval_test()
+    local dist_m_in = test_data[test_idx][1]
+    local signal_quality_pct_in = test_data[test_idx][2]
+    local status_expected = test_data[test_idx][3]
+
+    -- Use the client interface to get distance data and ensure it matches the distance data
+    -- that was sent through the driver interface.
+    local error_str = get_and_eval(test_idx, dist_m_in, signal_quality_pct_in, status_expected)
+    if error_str then
+        return complete(string.format("Test %i, dist_m: %.2f, quality_pct: %3i failed because %s",
+            test_idx, dist_m_in, signal_quality_pct_in, error_str))
+    end
+
+    -- Move to the next test in the list.
+    return update_begin_test()
+end
+
+update_prepare = _update_prepare
+update_wait = _update_wait
+update_begin_test = _update_begin_test
+update_eval_test = _update_eval_test
+
+send("Loaded rangefinder_quality_test.lua")
+
+return update_prepare, 0
```

Comment:
I think there should be a newline at the end? Not entirely sure

Reply:
I noticed that github flags this. Worth changing.